v0.34.0 fails on scripts using the dynamic keyword

### What You Are Seeing?
```C:\GitLab-Runner\builds\redacted>powershell -File build.ps1 -Verbosity Diagnostic
Preparing to run build script...
Running build script...
Module directory does not exist.
NuGet Config not specified. Will use NuGet default mechanism for resolving it.
Analyzing build script...
Analyzing C:/GitLab-Runner/builds/redacted/build.cake...
Processing build script...
Installing tools...
Installing addins...
The addin Cake.Figlet will reference Cake.Figlet.dll.
The addin Cake.FileHelpers will reference Cake.FileHelpers.dll.
The addin Cake.Incubator will reference Cake.Incubator.dll.
The addin Cake.ProGet will reference Cake.ProGet.dll.
Verifying assembly 'Cake.Figlet, Version=1.3.0.0, Culture=neutral, PublicKeyToken=null'.
Verifying assembly 'Cake.FileHelpers, Version=3.2.0.0, Culture=neutral, PublicKeyToken=null'.
Verifying assembly 'Cake.Incubator, Version=5.0.1.0, Culture=neutral, PublicKeyToken=null'.
Verifying assembly 'Cake.ProGet, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.
The assembly 'Cake.ProGet, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'
is referencing an older version of Cake.Core (0.28.1).
For best compatibility it should target Cake.Core version 0.33.0.
Adding assembly reference to mscorlib.dll...
Adding assembly reference to System.Core.dll...
Adding assembly reference to Cake.Core.dll...
Adding assembly reference to Cake.Common.dll...
Adding assembly reference to Cake.exe...
Adding assembly reference to System.dll...
Adding assembly reference to System.Xml.dll...
Adding assembly reference to System.Xml.Linq.dll...
Adding assembly reference to System.Data.dll...
Adding assembly reference to System.Runtime.dll...
Adding assembly reference to System.Collections.dll...
Adding assembly reference to System.Net.Http.dll...
Adding assembly reference to netstandard.dll...
Adding assembly reference to Cake.Figlet.dll...
Adding assembly reference to Cake.FileHelpers.dll...
Adding assembly reference to Cake.Incubator.dll...
Adding assembly reference to Cake.ProGet.dll...
Importing namespace Cake.Common...
Importing namespace Cake.Common.Build...
Importing namespace Cake.Common.Build.AppVeyor...
Importing namespace Cake.Common.Build.AppVeyor.Data...
Importing namespace Cake.Common.Build.Bamboo...
Importing namespace Cake.Common.Build.Bamboo.Data...
Importing namespace Cake.Common.Build.BitbucketPipelines...
Importing namespace Cake.Common.Build.BitbucketPipelines.Data...
Importing namespace Cake.Common.Build.Bitrise...
Importing namespace Cake.Common.Build.Bitrise.Data...
Importing namespace Cake.Common.Build.ContinuaCI...
Importing namespace Cake.Common.Build.ContinuaCI.Data...
Importing namespace Cake.Common.Build.GitLabCI...
Importing namespace Cake.Common.Build.GitLabCI.Data...
Importing namespace Cake.Common.Build.GoCD...
Importing namespace Cake.Common.Build.GoCD.Data...
Importing namespace Cake.Common.Build.Jenkins...
Importing namespace Cake.Common.Build.Jenkins.Data...
Importing namespace Cake.Common.Build.MyGet...
Importing namespace Cake.Common.Build.TeamCity...
Importing namespace Cake.Common.Build.TFBuild...
Importing namespace Cake.Common.Build.TFBuild.Data...
Importing namespace Cake.Common.Build.TravisCI...
Importing namespace Cake.Common.Build.TravisCI.Data...
Importing namespace Cake.Common.Diagnostics...
Importing namespace Cake.Common.IO...
Importing namespace Cake.Common.IO.Paths...
Importing namespace Cake.Common.Net...
Importing namespace Cake.Common.Security...
Importing namespace Cake.Common.Solution...
Importing namespace Cake.Common.Solution.Project...
Importing namespace Cake.Common.Solution.Project.Properties...
Importing namespace Cake.Common.Solution.Project.XmlDoc...
Importing namespace Cake.Common.Text...
Importing namespace Cake.Common.Tools...
Importing namespace Cake.Common.Tools.Cake...
Importing namespace Cake.Common.Tools.Chocolatey...
Importing namespace Cake.Common.Tools.Chocolatey.ApiKey...
Importing namespace Cake.Common.Tools.Chocolatey.Config...
Importing namespace Cake.Common.Tools.Chocolatey.Download...
Importing namespace Cake.Common.Tools.Chocolatey.Features...
Importing namespace Cake.Common.Tools.Chocolatey.Install...
Importing namespace Cake.Common.Tools.Chocolatey.New...
Importing namespace Cake.Common.Tools.Chocolatey.Pack...
Importing namespace Cake.Common.Tools.Chocolatey.Pin...
Importing namespace Cake.Common.Tools.Chocolatey.Push...
Importing namespace Cake.Common.Tools.Chocolatey.Sources...
Importing namespace Cake.Common.Tools.Chocolatey.Uninstall...
Importing namespace Cake.Common.Tools.Chocolatey.Upgrade...
Importing namespace Cake.Common.Tools.DotCover...
Importing namespace Cake.Common.Tools.DotCover.Analyse...
Importing namespace Cake.Common.Tools.DotCover.Cover...
Importing namespace Cake.Common.Tools.DotCover.Merge...
Importing namespace Cake.Common.Tools.DotCover.Report...
Importing namespace Cake.Common.Tools.DotNetCore...
Importing namespace Cake.Common.Tools.DotNetCore.Build...
Importing namespace Cake.Common.Tools.DotNetCore.BuildServer...
Importing namespace Cake.Common.Tools.DotNetCore.Clean...
Importing namespace Cake.Common.Tools.DotNetCore.Execute...
Importing namespace Cake.Common.Tools.DotNetCore.MSBuild...
Importing namespace Cake.Common.Tools.DotNetCore.NuGet.Delete...
Importing namespace Cake.Common.Tools.DotNetCore.NuGet.Push...
Importing namespace Cake.Common.Tools.DotNetCore.Pack...
Importing namespace Cake.Common.Tools.DotNetCore.Publish...
Importing namespace Cake.Common.Tools.DotNetCore.Restore...
Importing namespace Cake.Common.Tools.DotNetCore.Run...
Importing namespace Cake.Common.Tools.DotNetCore.Test...
Importing namespace Cake.Common.Tools.DotNetCore.Tool...
Importing namespace Cake.Common.Tools.DotNetCore.VSTest...
Importing namespace Cake.Common.Tools.DupFinder...
Importing namespace Cake.Common.Tools.Fixie...
Importing namespace Cake.Common.Tools.GitLink...
Importing namespace Cake.Common.Tools.GitReleaseManager...
Importing namespace Cake.Common.Tools.GitReleaseManager.AddAssets...
Importing namespace Cake.Common.Tools.GitReleaseManager.Close...
Importing namespace Cake.Common.Tools.GitReleaseManager.Create...
Importing namespace Cake.Common.Tools.GitReleaseManager.Export...
Importing namespace Cake.Common.Tools.GitReleaseManager.Label...
Importing namespace Cake.Common.Tools.GitReleaseManager.Publish...
Importing namespace Cake.Common.Tools.GitReleaseNotes...
Importing namespace Cake.Common.Tools.GitVersion...
Importing namespace Cake.Common.Tools.ILMerge...
Importing namespace Cake.Common.Tools.ILRepack...
Importing namespace Cake.Common.Tools.InnoSetup...
Importing namespace Cake.Common.Tools.InspectCode...
Importing namespace Cake.Common.Tools.MSBuild...
Importing namespace Cake.Common.Tools.MSpec...
Importing namespace Cake.Common.Tools.MSTest...
Importing namespace Cake.Common.Tools.NSIS...
Importing namespace Cake.Common.Tools.NuGet...
Importing namespace Cake.Common.Tools.NuGet.Add...
Importing namespace Cake.Common.Tools.NuGet.Init...
Importing namespace Cake.Common.Tools.NuGet.Install...
Importing namespace Cake.Common.Tools.NuGet.List...
Importing namespace Cake.Common.Tools.NuGet.Pack...
Importing namespace Cake.Common.Tools.NuGet.Push...
Importing namespace Cake.Common.Tools.NuGet.Restore...
Importing namespace Cake.Common.Tools.NuGet.SetApiKey...
Importing namespace Cake.Common.Tools.NuGet.SetProxy...
Importing namespace Cake.Common.Tools.NuGet.Sources...
Importing namespace Cake.Common.Tools.NuGet.Update...
Importing namespace Cake.Common.Tools.NUnit...
Importing namespace Cake.Common.Tools.OctopusDeploy...
Importing namespace Cake.Common.Tools.OpenCover...
Importing namespace Cake.Common.Tools.ReportGenerator...
Importing namespace Cake.Common.Tools.ReportUnit...
Importing namespace Cake.Common.Tools.Roundhouse...
Importing namespace Cake.Common.Tools.SignTool...
Importing namespace Cake.Common.Tools.SpecFlow...
Importing namespace Cake.Common.Tools.SpecFlow.StepDefinitionReport...
Importing namespace Cake.Common.Tools.SpecFlow.TestExecutionReport...
Importing namespace Cake.Common.Tools.TextTransform...
Importing namespace Cake.Common.Tools.VSTest...
Importing namespace Cake.Common.Tools.VSWhere...
Importing namespace Cake.Common.Tools.VSWhere.All...
Importing namespace Cake.Common.Tools.VSWhere.Latest...
Importing namespace Cake.Common.Tools.VSWhere.Legacy...
Importing namespace Cake.Common.Tools.VSWhere.Product...
Importing namespace Cake.Common.Tools.WiX...
Importing namespace Cake.Common.Tools.WiX.Heat...
Importing namespace Cake.Common.Tools.XBuild...
Importing namespace Cake.Common.Tools.XUnit...
Importing namespace Cake.Common.Xml...
Importing namespace Cake.Core...
Importing namespace Cake.Core.Diagnostics...
Importing namespace Cake.Core.IO...
Importing namespace Cake.Core.Scripting...
Importing namespace Cake.Figlet...
Importing namespace Cake.FileHelpers...
Importing namespace Cake.Incubator.AssertExtensions...
Importing namespace Cake.Incubator.DotNetBuildExtensions...
Importing namespace Cake.Incubator.DotNetCoreTestExtensions...
Importing namespace Cake.Incubator.EnumerableExtensions...
Importing namespace Cake.Incubator.FileExtensions...
Importing namespace Cake.Incubator.FilePathExtensions...
Importing namespace Cake.Incubator.FileSystemExtensions...
Importing namespace Cake.Incubator.GlobbingExtensions...
Importing namespace Cake.Incubator.LoggingExtensions...
Importing namespace Cake.Incubator.Project...
Importing namespace Cake.Incubator.ProjectPathExtensions...
Importing namespace Cake.Incubator.SolutionParserExtensions...
Importing namespace Cake.Incubator.StringExtensions...
Importing namespace Cake.ProGet.Asset...
Importing namespace Cake.ProGet.Universal...
Importing namespace Cake.ProGet.Universal.Install...
Importing namespace Cake.ProGet.Universal.Pack...
Importing namespace Cake.ProGet.Universal.Push...
Importing namespace Cake.ProGet.Universal.Unpack...
Importing namespace System...
Importing namespace System.Collections.Generic...
Importing namespace System.Dynamic...
Importing namespace System.IO...
Importing namespace System.Linq...
Importing namespace System.Text...
Importing namespace System.Threading.Tasks...
Compiling build script...
Error: Microsoft.CodeAnalysis.Scripting.CompilationErrorException: C:/GitLab-Runner/builds/redacted/build.cake(20,1): error CS0656: Missing compiler required member 'Microsoft.CSharp.RuntimeBinder.CSharpArgumentInfo.Create'
   at Microsoft.CodeAnalysis.Scripting.ScriptBuilder.ThrowIfAnyCompilationErrors(DiagnosticBag diagnostics, DiagnosticFormatter formatter)
   at Microsoft.CodeAnalysis.Scripting.ScriptBuilder.CreateExecutor[T](ScriptCompiler compiler, Compilation compilation, Boolean emitDebugInformation, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Scripting.Script`1.GetExecutor(CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Scripting.Script`1.RunAsync(Object globals, Func`2 catchException, CancellationToken cancellationToken)
   at Cake.Scripting.Roslyn.RoslynScriptSession.Execute(Script script)
   at Cake.Core.Scripting.ScriptRunner.Run(IScriptHost host, FilePath scriptPath, IDictionary`2 arguments)
   at Cake.Commands.BuildCommand.Execute(CakeOptions options)
   at Cake.CakeApplication.Run(CakeOptions options)
   at Cake.Program.Main()
```

### What is Expected?
```C:\GitLab-Runner\builds\redacted>powershell -File build.ps1 -Verbosity Diagnostic
Preparing to run build script...
<snip>
Time Elapsed 00:05:51.72
Finished executing task: Build
```

### What You Are Seeing?
Cake v0.34.0 fails on scripts that use the dynamic keyword

### What version of Cake are you using?
v0.34.0

### Are you running on a 32 or 64 bit system?
64 bit

### What environment are you running on?  Windows? Linux? Mac?
Windows

### Are you running on a CI Server?  If so, which one?
I'm seeing the same error when the build runs from our private gitlab runner and on my local machine.

### How Did You Get This To Happen? (Steps to Reproduce)
All builds using cake v0.34.0 suddenly started to fail.
Worked around the issue by pinning cake to v0.33.0.

--- a/.travis.yml
+++ b/.travis.yml
@@ -14,7 +14,7 @@ mono:
   - 5.12.0
   - 5.20.1
 
-dotnet: 2.1.603
+dotnet: 2.1.701
 
 before_install:
   - git fetch --unshallow # Travis always does a shallow clone, but GitVersion needs the full history including branches and tags

--- a/build.cake
+++ b/build.cake
@@ -633,9 +633,12 @@ Task("Create-Release-Notes")
 });
 
 Task("Prepare-Integration-Tests")
-    .IsDependentOn("Validate-Version")
+    .IsDependentOn("Create-NuGet-Packages")
     .Does(() =>
 {
+   Unzip(parameters.Paths.Directories.NuGetRoot.CombineWithFilePath($"Cake.Tool.{parameters.Version.SemVersion}.nupkg"),
+        parameters.Paths.Directories.IntegrationTestsBinTool);
+
    CopyDirectory(parameters.Paths.Directories.ArtifactsBinFullFx, parameters.Paths.Directories.IntegrationTestsBinFullFx);
    CopyDirectory(parameters.Paths.Directories.ArtifactsBinNetCore, parameters.Paths.Directories.IntegrationTestsBinNetCore);
 });
@@ -645,6 +648,7 @@ Task("Run-Integration-Tests")
     .DeferOnError()
     .DoesForEach(
         ()=> new[] {
+            GetFiles($"{parameters.Paths.Directories.IntegrationTestsBinTool.FullPath}/**/Cake.dll").Single(),
             parameters.Paths.Directories.IntegrationTestsBinFullFx.CombineWithFilePath("Cake.exe"),
             parameters.Paths.Directories.IntegrationTestsBinNetCore.CombineWithFilePath("Cake.dll")
         },

--- a/build/paths.cake
+++ b/build/paths.cake
@@ -37,6 +37,7 @@ public class BuildPaths
         var integrationTestsBin = context.MakeAbsolute(context.Directory("./tests/integration/tools"));
         var integrationTestsBinFullFx = integrationTestsBin.Combine("Cake");
         var integrationTestsBinNetCore = integrationTestsBin.Combine("Cake.CoreCLR");
+        var integrationTestsBinTool = integrationTestsBin.Combine("Cake.Tool");
 
         // Directories
         var buildDirectories = new BuildDirectories(
@@ -47,7 +48,8 @@ public class BuildPaths
             artifactsBinFullFx,
             artifactsBinNetCore,
             integrationTestsBinFullFx,
-            integrationTestsBinNetCore);
+            integrationTestsBinNetCore,
+            integrationTestsBinTool);
 
         // Files
         var buildFiles = new BuildFiles(
@@ -93,6 +95,7 @@ public class BuildDirectories
     public DirectoryPath ArtifactsBinNetCore { get; }
     public DirectoryPath IntegrationTestsBinFullFx { get; }
     public DirectoryPath IntegrationTestsBinNetCore { get; }
+    public DirectoryPath IntegrationTestsBinTool { get; }
     public ICollection<DirectoryPath> ToClean { get; }
 
     public BuildDirectories(
@@ -103,7 +106,8 @@ public class BuildDirectories
         DirectoryPath artifactsBinFullFx,
         DirectoryPath artifactsBinNetCore,
         DirectoryPath integrationTestsBinFullFx,
-        DirectoryPath integrationTestsBinNetCore
+        DirectoryPath integrationTestsBinNetCore,
+        DirectoryPath integrationTestsBinTool
         )
     {
         Artifacts = artifactsDir;
@@ -114,6 +118,7 @@ public class BuildDirectories
         ArtifactsBinNetCore = artifactsBinNetCore;
         IntegrationTestsBinFullFx = integrationTestsBinFullFx;
         IntegrationTestsBinNetCore = integrationTestsBinNetCore;
+        IntegrationTestsBinTool = integrationTestsBinTool;
         ToClean = new[] {
             Artifacts,
             TestResults,
@@ -122,7 +127,8 @@ public class BuildDirectories
             ArtifactsBinFullFx,
             ArtifactsBinNetCore,
             IntegrationTestsBinFullFx,
-            IntegrationTestsBinNetCore
+            IntegrationTestsBinNetCore,
+            IntegrationTestsBinTool
         };
     }
 }

--- a/src/Cake.Core/Cake.Core.csproj
+++ b/src/Cake.Core/Cake.Core.csproj
@@ -17,6 +17,7 @@
     <PackageReference Include="Microsoft.Win32.Registry" Version="4.4.0" />
     <PackageReference Include="Microsoft.Extensions.DependencyModel" Version="2.0.4" />
     <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
+    <PackageReference Include="Microsoft.CSharp" Version="4.5.0" />
   </ItemGroup>
   <!-- .NET Framework packages -->
   <ItemGroup Condition=" '$(TargetFramework)' == 'net46' ">

--- a/src/Cake.Core/Scripting/ScriptConventions.cs
+++ b/src/Cake.Core/Scripting/ScriptConventions.cs
@@ -62,6 +62,7 @@ namespace Cake.Core.Scripting
             var result = new HashSet<Assembly>(new SimpleAssemblyComparer());
             result.Add(typeof(Action).GetTypeInfo().Assembly); // mscorlib or System.Private.Core
             result.Add(typeof(IQueryable).GetTypeInfo().Assembly); // System.Core or System.Linq.Expressions
+            result.Add(typeof(Microsoft.CSharp.RuntimeBinder.CSharpArgumentInfo).Assembly); // Dynamic support
 
             // Load other Cake-related assemblies that we need.
             var cakeAssemblies = LoadCakeAssemblies(root);

--- a/src/Cake/Cake.csproj
+++ b/src/Cake/Cake.csproj
@@ -15,7 +15,7 @@
   </ItemGroup>
   <!-- Global packages -->
   <ItemGroup>
-    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Scripting" Version="3.0.0" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Scripting" Version="3.1.0" />
     <PackageReference Include="Autofac" Version="4.6.2" />
   </ItemGroup>
 </Project>

--- a/tests/integration/Cake.Core/Scripting/Dynamic.cake
+++ b/tests/integration/Cake.Core/Scripting/Dynamic.cake
@@ -1,0 +1,10 @@
+using System.Dynamic;
+
+Task("Cake.Core.Scripting.Dynamic")
+    .Does(() =>
+{
+    dynamic result = new ExpandoObject();
+    result.Result = true;
+
+    Assert.True(result.Result);
+});

--- a/tests/integration/build.cake
+++ b/tests/integration/build.cake
@@ -28,6 +28,7 @@
 #load "./Cake.Common/Tools/TextTransform/TextTransformAliases.cake"
 #load "./Cake.Core/Scripting/AddinDirective.cake"
 #load "./Cake.Core/Scripting/DefineDirective.cake"
+#load "./Cake.Core/Scripting/Dynamic.cake"
 #load "./Cake.Core/Scripting/HttpClient.cake"
 #load "./Cake.Core/Scripting/LoadDirective.cake"
 #load "./Cake.Core/Scripting/SystemCollections.cake"
@@ -48,6 +49,7 @@ var target = Argument<string>("target", "Run-All-Tests");
 Task("Cake.Core")
     .IsDependentOn("Cake.Core.Scripting.AddinDirective")
     .IsDependentOn("Cake.Core.Scripting.DefineDirective")
+    .IsDependentOn("Cake.Core.Scripting.Dynamic")
     .IsDependentOn("Cake.Core.Scripting.HttpClient")
     .IsDependentOn("Cake.Core.Scripting.LoadDirective")
     .IsDependentOn("Cake.Core.Scripting.SystemCollections")

